### PR TITLE
Shortcuts

### PR DESCRIPTION
**Pull request by vxsx**
Video of shortcuts in action http://d.pr/v/1bsDs

Screenshot of help modal (press ?):
screen shot 2016-08-15 at 17 19 54

### CODE REVIEW

File: cms/static/cms/js/modules/shortcuts/help.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    });
+
+    keyboard.setContext('cms');
+    keyboard.bind('?', function () {
+        modal.open({
+            title: CMS.config.lang.shortcuts,
+            width: 600,
+            height: 660,
+            html:
+                '<div class="cms-shortcuts">' +
```
Comment (FinalAngel) [inline cms/static/cms/js/modules/shortcuts/help.js]: any way to use a template here?

File: cms/static/cms/js/modules/shortcuts/placeholders.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
+var keyboard = require('keyboardjs');
+var $ = require('jquery');
+
+require('../jquery.trap');
+
+var bindPlaceholderKeys = function () {
```
Comment (FinalAngel) [inline cms/static/cms/js/modules/shortcuts/placeholders.js]: would be nice to have some comments describing what actions are bound, like ``registers [escape] shortcut for closing the structure focus``

File: cms/static/cms/js/modules/shortcuts/help.js (hunk 1 above)
Comment (vxsx) [inline cms/static/cms/js/modules/shortcuts/help.js]: was originally a template string that i later transformed with babel :D
yeah, would make more sense i think

File: cms/static/cms/js/modules/shortcuts/placeholders.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        var plugin = $('.cms-dragitem:focus');
+        var area = plugin.closest('.cms-dragarea');
+
+        $('.cms-dragitem').removeAttr('tabindex');
+        area.untrap();
+        area.focus();
+        keyboard.setContext('placeholders');
+    });
+
+    keyboard.bind('e', function () {
```
Comment (FinalAngel) [inline cms/static/cms/js/modules/shortcuts/placeholders.js]: same here, ``Press [e] to trigger the edit menu of a selected plugin`` or so

**Comment by FinalAngel** [comment]
very nice, +1 on screenshots :)

File: CHANGELOG.txt [hunk 4]
```diff
@@ -13,6 +13,7 @@
   structure and content mode.
 * Added ``Shift + Space`` shortcut that behaves similar to ``Space`` shortcut
   but takes into account currently hovered plugin.
+* Improved keyboard navigation
```
Comment (FinalAngel) [inline CHANGELOG.txt]: Also * Added information model about shortcuts

**Comment by FinalAngel** [comment]
LGTM !

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 5]
```diff
@@ -7,56 +7,121 @@
...
+        'confirm': '{% filter escapejs %}{% trans "Are you sure you want to delete this plugin?" %}{% endfilter %}',
+        'publish': '{% filter escapejs %}{% trans "Are you sure you want to publish this page?" %}{% endfilter %}',
+        'addPluginPlaceholder': '{% filter escapejs %}{% trans "Plugin will be added here" %}{% endfilter %}',
+        'confirmDirty': '{% filter escapejs %}{% trans "You have unsaved changes." %}{% endfilter %}',
+        'confirmDirtyESC': '{% filter escapejs %}{% trans "Are you sure you want to abandon these changes?" %}{% endfilter %}',
+        'errorLoadingEditForm': '{% filter escapejs %}{% trans "The form could not be loaded. Please check that the server is running correctly." %}{% endfilter %}',
+        'shortcuts': '{% filter escapejs %}{% trans "Shortcuts" %}{% endfilter %}',
+        'shortcutAreas': [
+            {
+                'title': 'CMS Wide Shortcuts',
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: _CMS-wide_

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 6]
```diff
@@ -7,56 +7,121 @@
...
+                'shortcuts': {
+                    'help': {
+                        'shortcut': '?',
+                        'desc': '{% filter escapejs %}{% trans "Bring up this help dialog" %}{% endfilter %}'
+                    },
+                    'esc': {
+                        'shortcut': 'ESC',
+                        'desc': '{% filter escapejs %}{% trans "Close / cancel" %}{% endfilter %}'
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: *Close/cancel* (no spaces)

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 7]
```diff
@@ -7,56 +7,121 @@
...
+                    },
+                    'toggle-structure-board': {
+                        'shortcut': 'space',
+                        'desc': '{% filter escapejs %}{% trans "Toggle structure mode" %}{% endfilter %}'
+                    },
+                    'toggle-structure-board-using-hovered-plugin': {
+                        'shortcut': 'shift+space',
+                        'desc': '{% filter escapejs %}{% trans "Toggle structure mode and highlight hovered plugin" %}{% endfilter %}'
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: Really should be *hovered-over plugin*, but even that is not very satisfactory. But still better than *hovered*.

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 8]
```diff
@@ -7,56 +7,121 @@
...
+                    },
+                    'create-dialog': {
+                        'shortcut': 'alt+c',
+                        'desc': '{% filter escapejs %}{% trans 'Open "Create" dialog' %}{% endfilter %}'
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: Can the word *Create* be in italics rather than quotes?

Comment (vxsx) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: i'd prefer not to have html in translations, but it is possible, yes

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 9]
```diff
@@ -7,56 +7,121 @@
...
+                    },
+                    'toolbar': {
+                        'shortcut': 'f > t / alt+t',
+                        'desc': '{% filter escapejs %}{% trans "Focus Toolbar" %}{% endfilter %}'
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: *Focus on Toolbar* - but do we want to commit to calling this *Toolbar* with a capital T?

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 10]
```diff
@@ -7,56 +7,121 @@
...
+                    },
+                }
+            },
+            {
+                'title': 'Structure Board',
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: Again we should decide on a consistent way of capitalising and stick to it.

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 11]
```diff
@@ -7,56 +7,121 @@
...
+                'shortcuts': {
+                    'placeholders': {
+                        'shortcut': 'f > p / alt+p',
+                        'desc': '{% filter escapejs %}{% trans "Focus placeholders" %}{% endfilter %}'
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: *Focus on placeholders*

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 12]
```diff
@@ -7,56 +7,121 @@
...
+                    },
+                    'traversing': {
+                        'shortcut': 'tab / shift+tab',
+                        'desc': '{% filter escapejs %}{% trans "Move to next / previous element" %}{% endfilter %}'
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: *next/previous*

File: cms/templates/cms/toolbar/toolbar_javascript.html (hunk 9 above)
Comment (vxsx) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: up to you. it's a major cms component so i figured it makes sense, maybe it's not :)

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 13]
```diff
@@ -7,56 +7,121 @@
...
+                        'shortcut': 'e',
+                        'desc': '{% filter escapejs %}{% trans "Edit plugin" %}{% endfilter %}'
+                    },
+                    'add': {
+                        'shortcut': '+ / a',
+                        'desc': '{% filter escapejs %}{% trans "Add plugin" %}{% endfilter %}'
+                    },
+                    'settings': {
+                        'shortcut': 's',
+                        'desc': '{% filter escapejs %}{% trans "Open settings dropdown" %}{% endfilter %}'
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: What is a settings dropdown?

File: cms/templates/cms/toolbar/toolbar_javascript.html [hunk 14]
```diff
@@ -7,56 +7,121 @@
...
+                    },
+                    'collapse': {
+                        'shortcut': 'x',
+                        'desc': '{% filter escapejs %}{% trans "Expand / collapse" %}{% endfilter %}'
```
Comment (evildmp) [inline cms/templates/cms/toolbar/toolbar_javascript.html]: *Expand/collapse*

**Pull request merged**